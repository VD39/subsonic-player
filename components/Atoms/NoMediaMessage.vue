<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="['column', 'centerItems', $style.content]">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: var(--width-height-100);
}

.content {
  position: absolute;
  inset: 50% auto auto 50%;
  transform: translate(-50%, -50%);
}
</style>
