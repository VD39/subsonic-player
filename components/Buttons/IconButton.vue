<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: string;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: number;
    iconWeight?: IconWeight;
    is?: ButtonType;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 24,
    iconWeight: 'duotone',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      $style.iconButton,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        [$style.fullWidth]: fullWidth,
      },
    ]"
    :disabled="disabled"
    :type="type"
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="phIcon"
      :class="$style.icon"
      aria-hidden="true"
      :size="iconSize"
      :weight="iconWeight"
      :color="iconColor"
    />

    <span
      ref="text"
      :class="[
        $style.text,
        {
          'visually-hidden': !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.iconButton {
  @mixin align-center;

  position: relative;
  display: inline-flex;
  gap: var(--space-8);
  padding: var(--button-spacing);
  margin: 0;
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.fullWidth {
  width: 100%;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  @mixin ellipsis-overflow;

  display: block;
  margin-bottom: 1px;
}
</style>
