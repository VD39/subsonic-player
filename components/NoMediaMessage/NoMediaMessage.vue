<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="$style.content">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: 100%;
}

.content {
  @mixin align-center;

  position: absolute;
  inset: 50% auto auto 50%;
  flex-direction: column;
  transform: translate(-50%, -50%);
}
</style>
