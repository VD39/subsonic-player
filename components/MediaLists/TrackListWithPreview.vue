<script setup lang="ts">
import TrackImage from '@/components/TrackDetails/TrackImage.vue';
import DropdownMenu from '@/components/Dropdown/DropdownMenu.vue';
import DropdownItem from '@/components/Dropdown/DropdownItem.vue';
import DropdownDivider from '@/components/Dropdown/DropdownDivider.vue';
import FavouriteButton from '@/components/Buttons/FavouriteButton.vue';
import ArtistsList from '@/components/TrackDetails/ArtistsList.vue';
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';

defineProps<{
  tracks: Track[];
}>();

defineEmits(['addToQueue', 'playTrack', 'removeFromPlaylist']);
</script>

<template>
  <div v-if="tracks.length" ref="tracksWrapper">
    <div class="trackHeader">
      <div class="trackPlay">#</div>
      <div class="trackDetails">
        <p>Track</p>
      </div>
      <div class="trackAlbum">
        <p>Album</p>
      </div>
      <div class="trackArtist">
        <p>Artists</p>
      </div>
      <div class="trackTime">
        <p>Time</p>
      </div>
    </div>

    <ol>
      <li
        v-for="(track, index) in tracks"
        :key="track.id"
        class="trackItem"
        data-test-id="track"
      >
        <div class="trackPlay">
          <TrackImage
            :track-id="track.id"
            @play-current-track="$emit('playTrack', track)"
          />
        </div>

        <div class="trackDetails">
          <p :class="['clamp', $style.name]">
            {{ track.trackNumber }}. {{ track.name }}
          </p>

          <FavouriteButton
            :id="track.id"
            type="album"
            :favourite="track.favourite"
          />

          <DropdownMenu data-test-id="dropdown-menu">
            <DropdownItem
              data-test-id="remove-from-playlist"
              @click="$emit('removeFromPlaylist', track)"
            >
              Remove from Playlist
            </DropdownItem>
            <DropdownDivider />
            <DropdownItem
              data-test-id="add-to-queue"
              @click="$emit('addToQueue', track)"
            >
              Add to queue
            </DropdownItem>
            <DropdownItem
              data-test-id="play-track"
              @click="$emit('playTrack', index)"
            >
              Play Track
            </DropdownItem>
          </DropdownMenu>
        </div>

        <div class="trackAlbum">
          <NuxtLink
            v-if="track.album && track.albumId"
            :to="`/album/${track.albumId}`"
            class="clamp"
          >
            {{ track.album }}
          </NuxtLink>
          <p v-else>{{ track.album }}</p>
        </div>

        <div class="trackArtist">
          <ArtistsList
            v-if="track.artists.length"
            :artists="track.artists"
            class="clamp"
          />
        </div>

        <time class="trackTime">
          {{ track.duration }}
        </time>
      </li>
    </ol>
  </div>
  <NoMediaMessage v-else icon="PhMusicNotes" message="No tracks found." />
</template>

<style module>
.name {
  margin: 0 auto 0 0;
}
</style>
