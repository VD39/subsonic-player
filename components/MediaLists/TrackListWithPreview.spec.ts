import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { getFormattedTracksMock } from '@/test/helpers';
import IconButton from '@/components/Buttons/IconButton.vue';
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';
import TrackListWithPreview from './TrackListWithPreview.vue';

function factory(props = {}) {
  return mount(TrackListWithPreview, {
    props: {
      tracks: [],
      ...props,
    },
  });
}

describe('TrackListWithPreview', () => {
  let wrapper: VueWrapper;

  beforeAll(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when radioStations prop is an empty array', () => {
    it('does not show the tracks wrapper', () => {
      expect(wrapper.find({ ref: 'tracksWrapper' }).exists()).toBe(false);
    });

    it('shows the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(true);
    });
  });

  describe('when radioStations prop is not an empty array', () => {
    const tracks = getFormattedTracksMock(5);

    beforeAll(async () => {
      wrapper = factory({
        tracks,
      });

      wrapper
        .findAllComponents('[data-test-id="dropdown-menu"]')[0]
        .findComponent(IconButton)
        .vm.$emit('click');

      await wrapper.vm.$nextTick();
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the tracks wrapper', () => {
      expect(wrapper.find({ ref: 'tracksWrapper' }).exists()).toBe(true);
    });

    it('shows the correct number of track items', () => {
      expect(wrapper.findAll('[data-test-id="track"]').length).toBe(5);
    });

    it('does not show the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(false);
    });

    describe('when the remove from playlist button is clicked', () => {
      beforeAll(() => {
        (
          wrapper.findAllComponents(
            '[data-test-id="remove-from-playlist"]',
          )[0] as VueWrapper
        ).vm.$emit('click');
      });

      it('emits the removeFromPlaylist event with track', () => {
        expect(wrapper.emitted('removeFromPlaylist')).toEqual([[tracks[0]]]);
      });
    });

    describe('when the add to queue button is clicked', () => {
      beforeAll(() => {
        (
          wrapper.findAllComponents(
            '[data-test-id="add-to-queue"]',
          )[0] as VueWrapper
        ).vm.$emit('click');
      });

      it('emits the addToQueue event with track', () => {
        expect(wrapper.emitted('addToQueue')).toEqual([[tracks[0]]]);
      });
    });

    describe('when the play track button is clicked', () => {
      beforeAll(() => {
        (
          wrapper.findAllComponents(
            '[data-test-id="play-track"]',
          )[0] as VueWrapper
        ).vm.$emit('click');
      });

      it('emits the playTrack event with track', () => {
        expect(wrapper.emitted('playTrack')).toEqual([[0]]);
      });
    });
  });
});
