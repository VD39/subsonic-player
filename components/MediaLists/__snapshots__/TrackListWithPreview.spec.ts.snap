// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrackListWithPreview > matches the snapshot 1`] = `
"<div class="noMediaMessage">
  <div class="content">
    <phmusicnotes size="50"></phmusicnotes>
    <p>No tracks found.</p>
  </div>
</div>"
`;

exports[`TrackListWithPreview > when radioStations prop is not an empty array > matches the snapshot 1`] = `
"<div>
  <div class="trackHeader">
    <div class="trackPlay">#</div>
    <div class="trackDetails">
      <p>Track</p>
    </div>
    <div class="trackAlbum">
      <p>Album</p>
    </div>
    <div class="trackArtist">
      <p>Artists</p>
    </div>
    <div class="trackTime">
      <p>Time</p>
    </div>
  </div>
  <ol>
    <li class="trackItem" data-test-id="track">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails">
        <p class="clamp name">. track-0</p><button class="iconButton" type="button" title="Like album">
          <phheart class="icon" aria-hidden="true" size="24" weight="duotone"></phheart><span class="text visually-hidden">Like album</span>
        </button>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <div class="dropdownMenu">
              <ul class="menu">
                <li class="dropdownItem" data-test-id="remove-from-playlist"><button class="iconButton fullWidth action action" type="button" data-test-id="remove-from-playlist">
                    <!--v-if--><span class="text"> Remove from Playlist </span>
                  </button></li>
                <li class="divider" aria-hidden="true"></li>
                <li class="dropdownItem" data-test-id="add-to-queue"><button class="iconButton fullWidth action action" type="button" data-test-id="add-to-queue">
                    <!--v-if--><span class="text"> Add to queue </span>
                  </button></li>
                <li class="dropdownItem" data-test-id="play-track"><button class="iconButton fullWidth action action" type="button" data-test-id="play-track">
                    <!--v-if--><span class="text"> Play Track </span>
                  </button></li>
              </ul>
            </div>
          </transition-stub>
        </div>
      </div>
      <div class="trackAlbum"><a class="clamp">album</a></div>
      <div class="trackArtist">
        <ul class="commaList clamp">
          <li data-test-id="artist-list-item"><a>name</a></li>
          <li data-test-id="artist-list-item"><a>name1</a></li>
        </ul>
      </div><time class="trackTime">00:19</time>
    </li>
    <li class="trackItem" data-test-id="track">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails">
        <p class="clamp name">. track-1</p><button class="iconButton" type="button" title="Like album">
          <phheart class="icon" aria-hidden="true" size="24" weight="duotone"></phheart><span class="text visually-hidden">Like album</span>
        </button>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
      <div class="trackAlbum"><a class="clamp">album</a></div>
      <div class="trackArtist">
        <ul class="commaList clamp">
          <li data-test-id="artist-list-item"><a>name</a></li>
          <li data-test-id="artist-list-item"><a>name1</a></li>
        </ul>
      </div><time class="trackTime">00:19</time>
    </li>
    <li class="trackItem" data-test-id="track">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails">
        <p class="clamp name">. track-2</p><button class="iconButton" type="button" title="Like album">
          <phheart class="icon" aria-hidden="true" size="24" weight="duotone"></phheart><span class="text visually-hidden">Like album</span>
        </button>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
      <div class="trackAlbum"><a class="clamp">album</a></div>
      <div class="trackArtist">
        <ul class="commaList clamp">
          <li data-test-id="artist-list-item"><a>name</a></li>
          <li data-test-id="artist-list-item"><a>name1</a></li>
        </ul>
      </div><time class="trackTime">00:19</time>
    </li>
    <li class="trackItem" data-test-id="track">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails">
        <p class="clamp name">. track-3</p><button class="iconButton" type="button" title="Like album">
          <phheart class="icon" aria-hidden="true" size="24" weight="duotone"></phheart><span class="text visually-hidden">Like album</span>
        </button>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
      <div class="trackAlbum"><a class="clamp">album</a></div>
      <div class="trackArtist">
        <ul class="commaList clamp">
          <li data-test-id="artist-list-item"><a>name</a></li>
          <li data-test-id="artist-list-item"><a>name1</a></li>
        </ul>
      </div><time class="trackTime">00:19</time>
    </li>
    <li class="trackItem" data-test-id="track">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails">
        <p class="clamp name">. track-4</p><button class="iconButton" type="button" title="Like album">
          <phheart class="icon" aria-hidden="true" size="24" weight="duotone"></phheart><span class="text visually-hidden">Like album</span>
        </button>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
      <div class="trackAlbum"><a class="clamp">album</a></div>
      <div class="trackArtist">
        <ul class="commaList clamp">
          <li data-test-id="artist-list-item"><a>name</a></li>
          <li data-test-id="artist-list-item"><a>name1</a></li>
        </ul>
      </div><time class="trackTime">00:19</time>
    </li>
  </ol>
</div>"
`;
